File: leetcode/python/703_kth_largest_element_stream.py
import heapq


class KthLargest:
    def __init__(self, k: int, nums: list[int]):
        self.k = k
        self.heap = nums
        heapq.heapify(self.heap)
        while len(self.heap) > k:
            heapq.heappop(self.heap)

    def add(self, val: int) -> int:
        heapq.heappush(self.heap, val)
        if len(self.heap) > self.k:
            heapq.heappop(self.heap)
        return self.heap[0]

t = KthLargest(3, [4, 5, 8, 2])
print(t.add(3))
print(t.add(5))
print(t.add(10))
print(t.add(9))
print(t.add(4))
        

File: leetcode/rust/src/703_kth_largest_element_stream.rs
use std::{collections::BinaryHeap, cmp::Reverse};

struct KthLargest {
    k: i32,
    heap: BinaryHeap<Reverse<i32>>,
}

impl KthLargest {
    fn new(k: i32, nums: Vec<i32>) -> Self {
        let mut heap = BinaryHeap::with_capacity(k as usize);
        for num in nums {
            heap.push(Reverse(num));
            if heap.len() > k as usize {
                heap.pop();
            }
        }
        Self { k, heap }
    }
    
    fn add(&mut self, val: i32) -> i32 {
        self.heap.push(Reverse(val));
        if self.heap.len() > self.k as usize {
            self.heap.pop();
        }

        let rev = self.heap.peek().unwrap();
        let Reverse(v) = *rev;
        v
    }
}
